src/Inciter/AMR/refinement.h

```diff
@@ -0,0 +1,779 @@
+
+                // This order should give a positive Jacobian
+                tet_store->add(child[0], A, AB, AC, AD, Refinement_Case::one_to_eight, tet_id);
+                tet_store->add(child[1], B, BC, AB, BD, Refinement_Case::one_to_eight, tet_id);
+                tet_store->add(child[2], C, AC, BC, CD, Refinement_Case::one_to_eight, tet_id);
+                tet_store->add(child[3], D, AD, CD, BD, Refinement_Case::one_to_eight, tet_id);
+
+                tet_store->add(child[4], BC, CD, AC, BD, Refinement_Case::one_to_eight, tet_id);
+                tet_store->add(child[5], AB, BD, AC, AD, Refinement_Case::one_to_eight, tet_id);
+                tet_store->add(child[6], AB, BC, AC, BD, Refinement_Case::one_to_eight, tet_id);
+                tet_store->add(child[7], AC, BD, CD, AD, Refinement_Case::one_to_eight, tet_id);
+
+                tet_store->deactivate(tet_id);
+
+            }
+
+            // This is just a simple assignment, but I wanted to abstract it
+            // for if we change the underlying type to something which a simple
+            // assignment is no longer safe
+            /**
+             * @brief Function to duplicate (deep copy) a tet. Useful for when
+             * you want to make a tet that's very similar to an existing one
+             *
+             * @param out The tet to store the copy
+             * @param original The tet to copy the data from
+             */
+            void copy_tet(tet_t* out, tet_t* original)
+            {
+                // NOTE: This will do a deep copy, so is safer than it may look
+                *out = *original;
+            }
+
+            // This is just a std::replace, but may need to be more complicated
+            // in the future?
+            /**
+             * @brief function to take an existing list of tet ids and
+             * replace one. This can be useful for when you want to build very
+             * similar tets which share nodes
+             *
+             * @param tet Tet to perform operation on
+             * @param remove Element to be replaced
+             * @param add Element to replace with
+             */
+            void replace_node(tet_t* tet, size_t remove, size_t add)
+            {
+                std::replace(tet->begin(), tet->end(), remove, add);
+            }
+
+            /**
+             * @brief Function to find out slot in the x,y,z data arrays a tet lives
+             *
+             * // NOTE: this is _currently_ trivial, but will be nice if we for
+             * example swap data stores to a map
+             *
+             * @param tet tet of the tet to look for
+             * @param element offset into that tet to look at
+             *
+             * @return tet into data arrays the tet lives
+             */
+            // TODO: Move this (or rename?)
+            size_t tet_id_to_node_id(size_t tet, size_t element) {
+                return tet_store->get(tet)[element];
+            }
```

Comment:
It would be good to put in an `Assert( element < NUM_TET_NODES, "Indexing out of tet )";` before `return`.

Also, just to double-check: is it the desired behavior that if `tet` does not exist in tet_store an exception is thrown? (I'm not sure, but you will know.)